### ISSUE TITLE
unlink of unix socket in connectionLost may cause twistd daemons to ignore SIGTERM

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3654]
|<img alt="jamesyonan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jamesyonan reported|
|-|-|
|Trac ID|trac#3654|
|Type|defect|
|Created|2009-02-12 06:35:51Z|
|Branch|https://github.com/twisted/twisted/tree/failed-connectionLost-3654|

In twisted/internet/unix.py, the Port.connectionLost method fails to catch exceptions arising from the os.unlink(self.port) statement.  The end result is that a twistd daemon that drops root privileges may not have sufficient privileges to unlink a unix socket on program shutdown.  This bug essentially causes twistd daemons that listen on unix domain sockets and drop root privileges to not be stoppable via a single SIGTERM because os.unlink throws an exception which halts the shutdown process.  A second SIGTERM must be delivered to stop the process.  The workaround is to bracket os.unlink with try/except, such as

        try:
            os.unlink(self.port) # this might raise an exception if privileges have been downgraded
        except:
            pass

<details><summary>Searchable metadata</summary>

```
trac-id__3654 3654
type__defect defect
reporter__jamesyonan jamesyonan
priority__highest highest
milestone__ 
branch__branches_failed_connectionLost_3654 branches/failed-connectionLost-3654
branch_author__washort washort
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1234420551000000 1234420551000000
changetime__1235832045000000 1235832045000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Another way to approach this is to consider that the reactor should perhaps be careful about calling the `connectionLost` method of selectables when it is shutting down.  Exceptions could be handled there, addressing any possible failure in any kind of filedescriptor.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_2"></a>|@washort commented|
|-|-|

(In [26244]) Branching to 'failed-connectionLost-3654'

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_3"></a>|@washort removed owner|
|-|-|

I can't reproduce this -- a trivial program listening on a root-owned unix socket deletes it successfully, even if its uid is changed.

However, I wrote a unit test for some uncovered code as part of my investigation. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @washort|
|-|-|

Deletion permission goes with containing directory permissions/ownership, not file permissions/ownership.  What happens if the directory the unix socket resides in is root-owned and not world-writable?

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_5"></a>|@washort removed owner|
|-|-|

Good point, for some reason I didn't remember that. Turns out the actual problem occurs when using `twisted.application.internet.UNIXServer` -- shutting down twistd causes the socket to be closed by service code. A failure in `connectionLost` prevents the Deferred returned by `loseConnection` from ever firing, thus preventing shutdown. I've changed it so that the Deferred returned by loseConnection has its errback fired upon a failure in `connectionLost`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Going to adjust some things here.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Okay, I think that's a bit simpler.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @exarkun|
|-|-|

This is a nice refactoring. The only thing I'm wondering is why `test_shutdownDisconnectsCleanly` is passing on trunk? It looks like the problem should be fixed by the branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

`test_shutdownDisconnectsCleanly` exercises a codepath which wasn't explicitly tested in trunk before.  It makes sure `reactor.disconnectAll` handles exceptions from `connectionLost`, which it did already.

The bug was in `tcp.Port.stopListening` (or, really, `unix.Port.stopListening` - it overrides `connectionLost` and introduces a place for this bug) which, while it _was_ raises an exception from `connectionLost`, the actual problem was that this caused it to fail to fire the `Deferred` returned by `stopListening`.  If you were using services, this would be a major problem, since `stopService` would then wait forever (or until the next SIGTERM).

The new `twisted.internet.test.test_posixbase.TCPPortTests.test_connectionLostFailed` tests the error case of `stopListening` and makes sure the `Deferred` still fires, so that's the one that needs this branch in order to pass.

It'd be really great if I had managed to make this clear through changes in the branch instead of having to explain it here, but I haven't figured out how to make this kind of "diff annotation" - where the code makes sense (I think), but the _change_ isn't as clear as it could be.  Oh well.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [26296]) Merge failed-connectionLost-3654 

Author: washort, exarkun
Reviewer: therve
Fixes: [#3654](https://github.com/twisted/twisted/issues/3654)

Take care to handle exceptions raised by `FileDescriptor.connectionLost` in
the delayed call to that method by `Port.stopListening`.  Previously an exception
would cause the `Deferred` returned to never fire; now it will errback with
the exception instead.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
